examples/async_api_check.py

```diff
@@ -0,0 +1,201 @@
+
+    async def fetch_gaze(self):
+        async for gaze in receive_gaze_data(
+            self.gaze_sensor.url, run_loop=True, log_level=30
+        ):
+            await self.gaze_queue.put(gaze)
+
+    async def start_streaming_task(self):
+        if self.stream_task:
+            self.stream_task.cancel()
+            self.stream_task = None
+        self.stream_task = asyncio.create_task(self.fetch_gaze())
+
+    async def on_sensor_connect(self):
+        logger.debug('Sensor connected.')
+        if not self.stream_task:
+            await self.start_streaming_task()
+            logger.info("Sensor was connected. Relay will be started.")
+
+    async def on_sensor_disconnect(self):
+        logger.warning("Sensor was disconnected.")
+        if self.stream_task:
+            self.stream_task.cancel()
+            self.stream_task = None
+
+    async def check_sensors(self):
+        if self.gaze_sensor.connected:
+            await self.on_sensor_connect()
+        else:
+            await self.on_sensor_disconnect()
+
+    async def on_update(self, component):
+        if isinstance(component, Sensor):
+            if component.sensor == 'gaze' and component.conn_type == 'DIRECT':
+                await self.update_status()
+                await self.check_sensors()
+
+    async def make_notifier(self):
+        self.notifier = StatusUpdateNotifier(
+            self.connected_device, callbacks=[self.on_update]
+        )
+
+    async def start_status_updates(self):
+        await self.make_notifier()
+        await self.notifier.receive_updates_start()
+
+    async def start_receiving(self):
+        await self.update_status()
+        await self.start_status_updates()
+
+    async def cleanup(self):
+        await self.notifier.receive_updates_stop()
+        await self.connected_device.close()
+        if self.stream_task:
+            self.stream_task.cancel()
+
+
+class Adapter:
+    def __init__(self, selected_device):
+        self.receiver = DataReceiver(selected_device)
+        self.publisher = pi_gaze_relay.PupilInvisibleGazeRelay()
+        self.timeout_check_task = None
+        self.receiver_to_publisher_task = None
+
+    async def push_to_publisher(self):
+        while True:
+            sample = await self.receiver.gaze_queue.get()
+            self.publisher.push_gaze_sample(sample)
+
+    async def cleanup(self):
+        await self.receiver.cleanup()
+        self.receiver_to_publisher_task.cancel()
+
+    async def check_timeout_on_receiver(self, timeout=60):
+        empty_time = 0
+        while empty_time < timeout:
+            if not self.receiver.stream_task:
+                logger.warning(
+                    f'Gaze sensor stopped streaming. Disconnecting in {timeout-empty_time} seconds'
+                )
+                empty_time += 1
+            else:
+                empty_time = 0
+            await asyncio.sleep(1)
+        await self.cleanup()
```

Comment:
I wonder if moving the timeout check into `push_to_publisher()`'s `gaze_queue.get()` might be more effective. `receiver.stream_task` might exist without it actually receiving data.

I also wonder if disconnecting after X seconds is actually helpful or better: What problem does it solve? One could also print the time since the last received datum, giving the user an idea on how long the disconnect is gong on already.